src/qspi.rs

```diff
@@ -0,0 +1,548 @@
+    ]
+}
+
+pub trait QspiExt {
+    fn bank1<T, PINS>(
+        self,
+        clocks: &CoreClocks,
+        _pins: PINS,
+        frequency: T,
+        prec: rec::Qspi,
+    ) -> Qspi
+    where
+        T: Into<Hertz>,
+        PINS: PinsBank1;
+
+    fn bank2<T, PINS>(
+        self,
+        clocks: &CoreClocks,
+        _pins: PINS,
+        frequency: T,
+        prec: rec::Qspi,
+    ) -> Qspi
+    where
+        T: Into<Hertz>,
+        PINS: PinsBank2;
+
+    fn qspi_unchecked<T>(
+        self,
+        clocks: &CoreClocks,
+        frequency: T,
+        bank: Bank,
+        prec: rec::Qspi,
+    ) -> Qspi
+    where
+        T: Into<Hertz>;
+}
+
+pub struct Qspi {
+    rb: stm32::QUADSPI,
+}
+
+impl Qspi {
+    pub fn bank1<T, PINS>(
+        regs: stm32::QUADSPI,
+        clocks: &CoreClocks,
+        _pins: PINS,
+        frequency: T,
+        prec: rec::Qspi,
+    ) -> Self
+    where
+        T: Into<Hertz>,
+        PINS: PinsBank1,
+    {
+        Self::qspi_unchecked(regs, clocks, frequency, prec, Bank::One)
+    }
+
+    pub fn bank2<T, PINS>(
+        regs: stm32::QUADSPI,
+        clocks: &CoreClocks,
+        _pins: PINS,
+        frequency: T,
+        prec: rec::Qspi,
+    ) -> Self
+    where
+        T: Into<Hertz>,
+        PINS: PinsBank2,
+    {
+        Self::qspi_unchecked(regs, clocks, frequency, prec, Bank::Two)
+    }
+
+    pub fn qspi_unchecked<T>(
+        regs: stm32::QUADSPI,
+        clocks: &CoreClocks,
+        frequency: T,
+        prec: rec::Qspi,
+        bank: Bank,
+    ) -> Self
+    where
+        T: Into<Hertz>,
+    {
+        prec.enable();
+
+        // Disable QUADSPI before configuring it.
+        regs.cr.write(|w| w.en().clear_bit());
+
+        let spi_kernel_ck = match Self::get_clock(clocks) {
+            Some(freq_hz) => freq_hz.0,
+            _ => panic!("QSPI kernel clock not running!"),
+        };
+
+        while regs.sr.read().busy().bit_is_set() {}
+
+        // Configure the FSIZE to maximum. It appears that even when addressing is not used, the
+        // flash size violation may still trigger.
+        regs.dcr.write(|w| unsafe { w.fsize().bits(0x1F) });
+
+        // Clear all pending flags.
+        regs.fcr.write(|w| {
+            w.ctof()
+                .set_bit()
+                .csmf()
+                .set_bit()
+                .ctcf()
+                .set_bit()
+                .ctef()
+                .set_bit()
+        });
+
+        // Configure the communication method for QSPI.
+        regs.ccr.write(|w| unsafe {
+            w.fmode()
+                .bits(0)
+                .dmode()
+                .bits(0b01)
+                .admode()
+                .bits(0b01)
+                .adsize()
+                .bits(0)
+                .imode()
+                .bits(0)
+                .dcyc()
+                .bits(0)
+        });
+
+        let spi_frequency = frequency.into().0;
+        let divisor = match spi_kernel_ck / spi_frequency {
```

Comment:
Elsewhere we round up the divisor, so that the resulting clock is less than or equal to specified. I think we need to do that here too.

So it would be `match (spi_kernel_ck + spi_frequency - 1) / spi_frequency`

You can also use [pattern binding](https://doc.rust-lang.org/stable/reference/patterns.html#identifier-patterns) to avoid repeating the calculation on the line below

Reply:
Ah, I was unaware about pattern binding! That's a nice feature.

Reply:
Fixed in 172a69b2eed35031ac014e7760bbc69c4eead8c6

Reply:
I think it does still need the `- 1` in the match arm, or in the bitfield assignment below. The register mapping is:

reg = 0; clk / 1
reg = 1; clk / 2
and so on

The bound can be `1..=256`

Reply:
Fixed the off-by-one and bounds in 93b7d3de890b31beff5885711376178d50a7a747